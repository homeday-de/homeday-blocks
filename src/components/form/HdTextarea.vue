<template>
  <TextFieldBase
    v-bind="$attrs"
    :name="name"
    :label="label"
    :icon="icon"
    :error="error"
    :helper="helper"
    :active="isActive"
    :filled="isFilled"
    :disabled="disabled"
    @clear-click="clearTextarea"
    @status-click="focusTextarea"
  >
    <textarea
      ref="textarea"
      :autocomplete="autocomplete"
      :value="value"
      :id="name"
      :name="name"
      :placeholder="placeholderAttr"
      :required="required"
      :autofocus="autofocus"
      :style="{ height }"
      :disabled="disabled"
      class="textarea"
      @focus="handleFocus"
      @blur="handleBlur"
      @input="handleInput"
    />
  </TextFieldBase>
</template>

<script>
import merge from 'lodash/merge';
import { getMessages } from 'homeday-blocks/src/lang';
import TextFieldBase from './TextFieldBase.vue';

export default {
  name: 'HdTextarea',
  components: {
    TextFieldBase,
  },
  props: {
    label: {
      type: String,
      default: '',
    },
    name: {
      type: String,
      required: true,
    },
    value: {
      type: [String, Number],
      default: '',
    },
    placeholder: {
      type: String,
      default: '',
    },
    required: {
      type: Boolean,
      default: false,
    },
    autocomplete: {
      type: String,
      default: 'on',
    },
    autofocus: {
      type: Boolean,
      default: false,
    },
    height: {
      type: String,
      default: '100px',
    },
    lang: {
      type: String,
      default: 'de',
    },
    texts: {
      type: Object,
      default: () => ({}),
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    icon: {
      type: String,
      default: '',
    },
  },
  data() {
    return {
      isActive: undefined,
      isValid: undefined,
      error: null,
      helper: null,
    };
  },
  computed: {
    t() {
      return merge(getMessages(this.lang), this.texts);
    },
    isFilled() {
      return this.value !== null && this.value !== undefined && this.value !== '';
    },
    placeholderAttr() {
      if (!this.placeholder) {
        return '';
      }

      if (!this.label || this.isActive) {
        return this.placeholder;
      }

      return '';
    },
  },
  watch: {
    value() {
      this.validate();
    },
  },
  methods: {
    clearTextarea() {
      this.focusTextarea();
      this.$emit('input', '');
    },
    focusTextarea() {
      this.$refs.textarea.focus();
    },
    handleFocus() {
      this.isActive = true;
    },
    handleBlur() {
      this.isActive = false;
      this.validate();
    },
    handleInput(e) {
      this.$emit('input', e.target.value);
    },
    showError(errorMessage) {
      this.error = errorMessage;
      this.isValid = false;
    },
    showHelper(message) {
      this.helper = message;
    },
    hideError() {
      this.isValid = true;
      this.error = null;
    },
    validate() {
      if (this.required && !this.isFilled) {
        this.showError(this.t.FORM.VALIDATION.REQUIRED);
      } else {
        this.hideError();
      }
      return !this.error;
    },
  },
};
</script>

<style scoped lang="scss">
.textarea {
  display: block;
  resize: vertical;
}
</style>
