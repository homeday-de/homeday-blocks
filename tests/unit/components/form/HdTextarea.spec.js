import { wrapperFactoryBuilder } from 'tests/unit/helpers';
import FIELD_CLASSES from './FIELD_CLASSES';
import HdTextarea from '@/components/form/HdTextarea.vue';

const ERROR_SELECTOR = '.field__error';
const HELPER_SELECTOR = '.field__error--helper';
const TEST_VALUE = 'new value';
const ICON_SELECTOR = '.field__icon';

const wrapperBuilder = wrapperFactoryBuilder(HdTextarea, {
  props: {
    name: 'test name',
    label: 'test label',
    placeholder: 'test placeholder',
  },
});

describe('HdTextarea', () => {
  it('is rendered as expected', () => {
    const wrapper = wrapperBuilder();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('is rendered with prefilled value', () => {
    const wrapper = wrapperBuilder({
      props: {
        value: TEST_VALUE,
      },
    });

    expect(wrapper.find('textarea').element.value).toBe(TEST_VALUE);
  });

  it('emits `input` event with the right payload', () => {
    const wrapper = wrapperBuilder();

    wrapper.get('textarea').setValue(TEST_VALUE);

    expect(wrapper.emitted('input')[0][0]).toEqual(TEST_VALUE);
  });

  it('allows setting a helper', async () => {
    const wrapper = wrapperBuilder();
    const TEST_HTML = '<b>test</b> test';

    wrapper.vm.showHelper(TEST_HTML);

    await wrapper.vm.$nextTick();

    expect(wrapper.find(HELPER_SELECTOR).element.innerHTML).toEqual(TEST_HTML);
  });

  it('validates requiredness', async () => {
    const wrapper = wrapperBuilder({
      props: {
        value: '',
        required: true,
      },
    });

    expect(wrapper.vm.validate()).toBe(false);

    await wrapper.vm.$nextTick();

    expect(wrapper.find(ERROR_SELECTOR).text()).toBeTruthy();
    expect(wrapper.classes()).toContain(FIELD_CLASSES.INVALID);

    wrapper.setProps({
      required: false,
    });

    expect(wrapper.vm.validate()).toBe(true);
  });

  it('Supports disabling', () => {
    const wrapper = wrapperBuilder({
      props: {
        disabled: true,
      },
    });

    expect(wrapper.find('textarea').attributes().disabled).toBe('disabled');
  });

  it('Can render an icon', async () => {
    const wrapper = wrapperBuilder();
    const ICON_PATH = 'fake/icon.svg';

    expect(wrapper.classes()).not.toContain(FIELD_CLASSES.HAS_ICON);
    expect(wrapper.find(ICON_SELECTOR).exists()).toBe(false);

    wrapper.setProps({
      icon: ICON_PATH,
    });

    await wrapper.vm.$nextTick();

    expect(wrapper.classes()).toContain(FIELD_CLASSES.HAS_ICON);
    expect(wrapper.find(ICON_SELECTOR).exists()).toBe(true);
  });
});
