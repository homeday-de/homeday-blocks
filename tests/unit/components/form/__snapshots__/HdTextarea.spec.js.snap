// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdTextarea is rendered as expected 1`] = `
<div class="field text-field">
  <!---->
  <div class="field__body">
    <div class="field__main"><textarea autocomplete="on" id="test name" name="test name" placeholder="" class="textarea" style="height: 100px;"></textarea> <label for="test name" class="field__label">
        test label
      </label>
      <div class="field__input-right"></div>
      <div class="field__border"></div>
    </div>
    <p class="field__helper">‚Äç</p>
  </div>
</div>
`;
